File: android-sdui/app/src/main/java/com/hsbc/sdui/analytics/ObservabilityClient.kt
package com.hsbc.sdui.analytics

import android.os.SystemClock
import java.security.SecureRandom

object ObservabilityClient {
    private val random = SecureRandom()
    private var traceId: String = randomHex(16)
    private var appStartMs: Long = SystemClock.elapsedRealtime()
    private var foregroundStartMs: Long = appStartMs
    private var startupType: String = "cold"

    fun markAppCreate() {
        traceId = randomHex(16)
        appStartMs = SystemClock.elapsedRealtime()
        foregroundStartMs = appStartMs
        startupType = "cold"
        recordStartupStep("activity_on_create", 0)
    }

    fun markForeground() {
        foregroundStartMs = SystemClock.elapsedRealtime()
        startupType = "warm"
        recordStartupStep("activity_on_start", 0)
    }

    fun traceparent(): String = "00-$traceId-${randomHex(8)}-01"

    fun startupElapsedMs(): Long {
        val start = if (startupType == "warm") foregroundStartMs else appStartMs
        return SystemClock.elapsedRealtime() - start
    }

    fun recordStartupStep(step: String, durationMs: Long, screenId: String = "home-hub-hk") {
        AppDynamicsClient.reportStartupStep(
            name = step,
            durationMs = durationMs,
            traceId = traceId,
            spanId = randomHex(8),
            startupType = startupType,
            screenId = screenId
        )
    }

    fun recordNetworkStep(name: String, durationMs: Long, path: String, success: Boolean) {
        AppDynamicsClient.reportNetworkStep(
            name = name,
            durationMs = durationMs,
            traceId = traceId,
            spanId = randomHex(8),
            path = path,
            success = success
        )
    }

    private fun randomHex(byteCount: Int): String {
        val bytes = ByteArray(byteCount)
        random.nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

File: ios-sdui/HSBCSDUI/Analytics/ObservabilityClient.swift
import Foundation
import Security
import UIKit

public final class ObservabilityClient {
    public static let shared = ObservabilityClient()

    private var traceId = ObservabilityClient.randomHex(byteCount: 16)
    private var appStart = Date()
    private var foregroundStart = Date()
    private var startupType = "cold"

    private init() {}

    public func markAppStart() {
        traceId = Self.randomHex(byteCount: 16)
        appStart = Date()
        foregroundStart = appStart
        startupType = "cold"
        recordStartupStep("app_init", durationMs: 0)
    }

    public func markForeground() {
        foregroundStart = Date()
        startupType = "warm"
        recordStartupStep("scene_foreground", durationMs: 0)
    }

    public func traceparent() -> String {
        "00-\(traceId)-\(Self.randomHex(byteCount: 8))-01"
    }

    public func startupElapsedMs() -> Int {
        let start = startupType == "warm" ? foregroundStart : appStart
        return Int(Date().timeIntervalSince(start) * 1000)
    }

    public func recordStartupStep(_ step: String, durationMs: Int, screenId: String = "home-hub-hk") {
        AppDynamicsClient.shared.reportStartupStep(
            name: step,
            durationMs: durationMs,
            traceId: traceId,
            spanId: Self.randomHex(byteCount: 8),
            startupType: startupType,
            screenId: screenId
        )
    }

    public func recordNetworkStep(_ name: String, durationMs: Int, path: String, success: Bool) {
        AppDynamicsClient.shared.reportNetworkStep(
            name: name,
            durationMs: durationMs,
            traceId: traceId,
            spanId: Self.randomHex(byteCount: 8),
            path: path,
            success: success
        )
    }

    private static func randomHex(byteCount: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        _ = SecRandomCopyBytes(kSecRandomDefault, byteCount, &bytes)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
